### PR TITLE
feat(experiments): Add holdout groups

### PR DESCRIPTION
**Pull request by neilkakkar**
## Problem

Add holdout groups model and related flag updates

## Changes

👉 _Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review._

### CODE REVIEW

File: posthog/models/feature_flag/flag_matching.py [hunk 1]
```diff
@@ -334,7 +334,9 @@ def is_holdout_condition_match(self, feature_flag: FeatureFlag) -> tuple[bool, s
                 # rollout_percentage is None (=100%), or we are inside holdout rollout bound.
                 # Thus, we match. Now get the variant override for the holdout condition.
                 variant_override = condition.get("variant")
-                if variant_override in [variant["key"] for variant in feature_flag.variants]:
+                if variant_override:
+                    # TODO: Should the variant be validated against the variants in the flag here?
```
Comment (neilkakkar) [inline posthog/models/feature_flag/flag_matching.py:338]: doesn't seem worth it, good enough as is I think.

File: ee/clickhouse/views/experiment_holdouts.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    def _get_filters_with_holdout_id(self, id: int, filters: list) -> list:
+        variant_name = f"holdout-{id}"
+        updated_filters = []
+        for filter in filters:
+            updated_filters.append(
+                {
+                    **filter,
+                    "variant": variant_name,
+                }
+            )
```
Comment (jurajmajerik) [inline ee/clickhouse/views/experiment_holdouts.py:45]:
```suggestion
        variant_key = f"holdout-{id}"
        updated_filters = []
        for filter in filters:
            updated_filters.append(
                {
                    **filter,
                    "variant": variant_key,
                }
            )
```

File: posthog/models/experiment.py [hunk 3]
```diff
@@ -30,6 +30,8 @@ class ExperimentType(models.TextChoices):
     created_by = models.ForeignKey("User", on_delete=models.SET_NULL, null=True)
     feature_flag = models.ForeignKey("FeatureFlag", blank=False, on_delete=models.RESTRICT)
     exposure_cohort = models.ForeignKey("Cohort", on_delete=models.SET_NULL, null=True)
+    holdout = models.ForeignKey("ExperimentHoldout", on_delete=models.SET_NULL, null=True)
```
Comment (jurajmajerik) [inline posthog/models/experiment.py:33]: This references the entire model in the backend, but only returns the ID to the frontend (is this a serializer thing?).

Could we return the entire holdout instance in `experiment.holdout` like we do for e.g `experiment.feature_flag`? Would be less confusing and could grab the holdout name for the UI directly from there.

**Comment by daibhin** [comment]
Are ya'll planning on merging this soon (today)? I have another migration that I want to get in but don't want to block you with a migration. I can hold off if necessary

**Comment by jurajmajerik** [comment]
@daibhin yes merging soon!

File: posthog/models/experiment.py (hunk 3 above)
Comment (neilkakkar) [reply posthog/models/experiment.py:33]: done, indeed a serializer thing

Comment (jurajmajerik) [reply posthog/models/experiment.py:33]: okay let's merge this

**Review by jurajmajerik — APPROVED** [review]
🚢

**Comment by neilkakkar** [comment]
experiments table is small enough that the constraint creation shouldn't be a problem

**Comment by neilkakkar** [comment]
will merge flag-holdouts first and then this 👍

**Pull request merged**